### PR TITLE
[vcpkg] Add a schema file for vcpkg.json

### PR DESCRIPTION
**Pull request by edhebi**
This is related to #12175 .

This is absolutely not as tested as it should be, but I'm not sure how to go about it. I gess some set of known valid and known invalid manifest files would be required.

I choose the path an filename fairly arbitrarily, feel free to move it around.

### CODE REVIEW

File: scripts/vcpkg.schema.json [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      "oneOf": [
+        {
+          "$ref": "#/definitions/packageName"
+        },
+        {
+          "$ref": "#/definitions/dependencyObject"
+        }
+      ]
+    },
+    "platformExpression": {
```
Comment (Edhebi) [inline scripts/vcpkg.schema.json:82]: There's not check on the string content, as I don't think it's doable regexes

Comment (strega-nil) [reply scripts/vcpkg.schema.json:82]: I don't think it's possible since it's a CFG, and CFGs are strictly more powerful than regular languages.

Comment (Edhebi) [reply scripts/vcpkg.schema.json:82]: Yeah. I still think there's value in having a named definition here, just for the sake of self documentation ?

Comment (strega-nil) [reply scripts/vcpkg.schema.json:82]: It might be possible to do some basic checking of tokens, at least, but no it's not necessary.

**Comment by Edhebi** [comment]
Also, if that path ends up being the final one, an id field should be added with and url providing the schema, either the github raw one, or something else:
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id: "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json",
  // ...
}
```

File: scripts/vcpkg.schema.json (hunk 1 above)
Comment (Edhebi) [reply scripts/vcpkg.schema.json:82]: Fair enough. Do you want me to remove it ?

Comment (strega-nil) [reply scripts/vcpkg.schema.json:82]: why would you remove it?

**Comment by strega-nil** [comment]
I think that this is really good, but can you make the identifiers for the definitions match the spec? `platformExpression` -> `platform-expression`?

File: scripts/vcpkg.schema.json (hunk 1 above)
Comment (Edhebi) [reply scripts/vcpkg.schema.json:82]: the platformExpression definition, and just use `{ "type": "string" }` where it's used instead

**Comment by Edhebi** [comment]
I wondered wether too do that or use the conventional json casing. I'll change the case when I have some time

**Comment by Edhebi** [comment]
done, all definitions are now with-dashes. I kept description as `description-field` to avoid mixing it up with the `description` keyword

File: scripts/vcpkg.schema.json (hunk 1 above)
Comment (strega-nil) [reply scripts/vcpkg.schema.json]: No, I think this is better; more descriptive.

**Review by ras0219 — APPROVED** [review]

**Comment by ras0219** [comment]
@Edhebi did you intend to mark this as ready to merge instead of a draft?

**Comment by Edhebi** [comment]
@ras0219 oops, my bad. I'll mark it as ready to merge

File: scripts/vcpkg.schema.json [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+      "description": "Name of a package.",
+      "allOf": [
+        {
+          "description": "Package name must be a dot-separated list of valid identifiers",
+          "type": "string",
+          "pattern": "^[a-z0-9]+(-[a-z0-9]+)*(\\.[a-z0-9]+(-[a-z0-9]+)*)*$"
+        },
+        {
+          "not": {
+            "description": "Identifiers must not be a Windows filesystem reserved name.",
```
Comment (strega-nil) [inline scripts/vcpkg.schema.json:34]:
```suggestion
            "description": "Identifiers must not be a Windows filesystem or vcpkg reserved name.",
```

File: scripts/vcpkg.schema.json [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+      "description": "Identifiers used for feature names.",
+      "allOf": [
+        {
+          "description": "Identifier are lowercase with digits and dashes.",
+          "type": "string",
+          "pattern": "[a-z0-9]+(-[a-z0-9]+)*"
+        },
+        {
+          "not": {
+            "description": "Identifiers must not be a Windows filesystem reserved name.",
```
Comment (strega-nil) [inline scripts/vcpkg.schema.json:17]:
```suggestion
            "description": "Identifiers must not be a Windows filesystem or vcpkg reserved name.",
```

File: scripts/vcpkg.schema.json [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        "dependencies": {
+          "description": "Dependencies used by this feature.",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/dependency"
+          }
+        }
+      },
+      "patternProperties": {
+        "^[$]": {}
```
Comment (strega-nil) [inline scripts/vcpkg.schema.json:118]:
```suggestion
        "^\\$": {}
```

**Review by strega-nil — APPROVED** [review]

**Comment by strega-nil** [comment]
Thanks so much @Edhebi!

**Pull request merged**